### ISSUE TITLE
Is it possible to query and filter fields of JSONB columns?

### ISSUE THREAD

**Issue by barbalex** [postgraphql/postgraphql#642]
I'm submitting a ... 
- [x] question

PostGraphQL version: 4.0.0-alpha2.25 (which by the way works like a charm)

I need to filter data based on fields contained in JSONB columns.

These fields have variable structure which is why they are saved in JSONB columns.

So far I have seen this possibility:
1. Build function(s) that accept Arrays of conditions: `key`, `value`, `comparator` for filtering and an array of keys to choose what fields to return
2. These Functions would return the data using PostgreSQL JSONB querying tools

This has some marked downsides:
- I need to write complex code in a language I do not know well (pgsql). I would much rather do this in JavaScript
- The data will not be preserved in my Apollo v2 store as it is returned with a different structure (columns for the JSON Fields)

Is there a better solution? My dream would be to be able to query these fields directly in graphql as if they were any other regular columns.

**Comment by benjie** [comment]
Take a look at https://github.com/mattbretl/postgraphile-filter-plugins as a starting point - you can extend the filters with JSONB tooling.

Also check the docs on graphile.org if you haven’t already.

**Comment by benjie** [comment]
(It’s still quite verbose writing plugins right now, the system needs to mature a bit and gain some tooling.)

JSONB is problematic as regards auto-discovery because how can we know what fields to export automatically? So either you need to write a custom solution (e.g. plugin) OR we need to figure out a way to indicate the columns you want. This seems like a perfect use case for smart comments - have a search through the other issues for smart comments, there’s some work in progress on them. We may want to reconsider the spec slightly; ai can’t remember if it lets you use the same field multiple times or not. I’m thinking

```
@jsonField date timestamp
@jsonField name text
@jsonField episode enum ONE=1 TWO=2
```

Or something like that.

Might be worth looking at massivejs for inspiration, I think Dian has some good querying stuff in place for JSONB?

G’night! 💤

**Comment by barbalex** [comment]
@benjie Thanks a lot, that is very helpful. And also: congratulations for making plugins possible!

**Comment by benjie** [comment]
No problem; feel free to keep raising issues and if your needs are more conversational do check out our Gitter:

https://gitter.im/postgraphql/postgraphql

**Comment by benjie** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [#473 - magic comments / deprecate fields](https://trello.com/c/Cv1bxDeG/19-473-magic-comments-deprecate-fields)

**Comment by git-blame** [comment]
Although there are commits for this issue, I've searched through postgraphile documentation and I don't see any examples or descriptions.

Is it possible to return a specific field of a JSONB column by creating a "jsonField" smart comment or tag?

**Comment by benjie** [comment]
Only if you write a plugin to do so. You might find the JSONStructure example relevant: https://github.com/graphile/cookbook/tree/master/examples/json_structure

**Comment by andrewrduncan** [comment]
I just had to do this and it was a Long Journey to a short solution. I haven't had time to put this into a sample project or anything as I'm actively in development on client effort where I needed this. Here's quick and dirty. My solution is very specific to Dictionary<,> in jsonb columns. Think attributes.

1. I created a CustomFilterConvention
2. Within that convention I register the defaults
3. I add in any new Operations I may need that they don't already account for
4. I then scan the provided DbContext and automatically look for all IDictionary types on the Schema and register them for my custom DictionaryFilterType
5. Form there I register my custom handlers
6. On going enhancement will be to add in additional operations and handlers as needed.

Sorry for the long post. I hope this is helpful to someone out there.

To use drop in place and call:
.AddFiltering<CustomQueryableFilterConvention<MyDbContext>>()

[Filtering.zip](https://github.com/graphile/postgraphile/files/9993384/Filtering.zip)
